File: apps/server-admin/src/lib/ddns/providers/tencentcloud.ts
import type {
  DDNSProviderContext,
  DDNSProviderDefinition,
  DDNSUpdateResult,
} from "../types";
import {
  ddnsProviderT,
  requestTencentCloudJson,
  splitDomain,
  toPositiveInt,
  updateDualStack,
} from "./helpers";

const TENCENTCLOUD_DNSPOD_HOST = "dnspod.tencentcloudapi.com";
const TENCENTCLOUD_DNSPOD_SERVICE = "dnspod";
const TENCENTCLOUD_DNSPOD_VERSION = "2021-03-23";
const TENCENTCLOUD_DEFAULT_LINE = "\u9ed8\u8ba4";
const tencentcloudT = (
  key: string,
  params?: Record<string, string | number | boolean | null | undefined>,
) => ddnsProviderT("tencentcloud", key, params);
const commonT = (
  key: string,
  params?: Record<string, string | number | boolean | null | undefined>,
) => ddnsProviderT("common", key, params);

type TencentCloudDescribeRecordListResponse = {
  RecordList?: Array<{
    Name?: string;
    RecordId: number;
    Type?: string;
    Value?: string;
    Line?: string;
    LineId?: string;
    TTL?: number;
  }>;
};

type TencentCloudRecordChangeResponse = {
  RecordId?: number;
};

export const tencentcloudProvider: DDNSProviderDefinition = {
  name: "tencentcloud",
  label: tencentcloudT("label"),
  fields: [
    {
      key: "secret_id",
      label: "SecretId",
      type: "text",
      placeholder: "AKID...",
      required: true,
    },
    {
      key: "secret_key",
      label: "SecretKey",
      type: "password",
      placeholder: tencentcloudT("fields.secret_key.placeholder"),
      required: true,
    },
    {
      key: "root_domain",
      label: commonT("fields.root_domain.label"),
      type: "text",
      placeholder: "example.com",
      required: true,
      description: commonT("fields.root_domain.description"),
    },
    {
      key: "domain",
      label: commonT("fields.domain.label"),
      type: "text",
      placeholder: "home.example.com",
      required: true,
      description: commonT("fields.domain.hostDescription"),
    },
    {
      key: "record_line",
      label: tencentcloudT("fields.record_line.label"),
      type: "text",
      placeholder: tencentcloudT("defaultLine"),
      required: false,
      description: tencentcloudT("fields.record_line.description"),
    },
    {
      key: "record_line_id",
      label: tencentcloudT("fields.record_line_id.label"),
      type: "text",
      placeholder: "0",
      required: false,
      description: tencentcloudT("fields.record_line_id.description"),
    },
    {
      key: "ttl",
      label: "TTL",
      type: "text",
      placeholder: "600",
      required: false,
      description: commonT("fields.ttl.description", { seconds: 600 }),
    },
  ],
};

async function tencentcloudRequest<T>(
  context: DDNSProviderContext,
  action: string,
  payload: Record<string, unknown>,
): Promise<T> {
  const { config, http } = context;
  const secretId = config.secret_id?.trim();
  const secretKey = config.secret_key?.trim();
  if (!secretId || !secretKey) {
    throw new Error(tencentcloudT("configIncomplete"));
  }

  return requestTencentCloudJson<T>(http, {
    action,
    host: TENCENTCLOUD_DNSPOD_HOST,
    payload,
    secretId,
    secretKey,
    service: TENCENTCLOUD_DNSPOD_SERVICE,
    version: TENCENTCLOUD_DNSPOD_VERSION,
  });
}

async function describeTencentCloudRecordList(
  context: DDNSProviderContext,
  payload: Record<string, unknown>,
): Promise<TencentCloudDescribeRecordListResponse> {
  try {
    return await tencentcloudRequest<TencentCloudDescribeRecordListResponse>(
      context,
      "DescribeRecordList",
      payload,
    );
  } catch (error) {
    const message = error instanceof Error ? error.message : String(error);
    if (message.startsWith("ResourceNotFound.NoDataOfRecord:")) {
      return { RecordList: [] };
    }
    throw error;
  }
}

export async function tencentcloudUpdate(
  context: DDNSProviderContext,
  ipv4: string | null,
  ipv6: string | null,
): Promise<DDNSUpdateResult> {
  const { config } = context;
  const secretId = config.secret_id?.trim();
  const secretKey = config.secret_key?.trim();
  const rootDomain = config.root_domain?.trim();
  const domain = config.domain?.trim();
  if (!secretId || !secretKey || !rootDomain || !domain) {
    return { success: false, message: tencentcloudT("configIncomplete") };
  }

  const ttl = toPositiveInt(config.ttl, 600);
  const parsed = splitDomain(domain, rootDomain);
  const recordLine = config.record_line?.trim() || TENCENTCLOUD_DEFAULT_LINE;
  const recordLineId = config.record_line_id?.trim();

  return updateDualStack(tencentcloudT("label"), ipv4, ipv6, async (recordType, ip) => {
    const basePayload: Record<string, unknown> = {
      Domain: parsed.rootDomain,
      RecordType: recordType,
    };

    if (recordLineId) {
      basePayload.RecordLineId = recordLineId;
    } else {
      basePayload.RecordLine = recordLine;
    }

    const list = await describeTencentCloudRecordList(context, {
      ...basePayload,
      Limit: 100,
      Offset: 0,
      Subdomain: parsed.recordName,
    });

    const existing = (list.RecordList || []).find((record) => {
      if ((record.Name || parsed.recordName) !== parsed.recordName) {
        return false;
      }
      if ((record.Type || recordType) !== recordType) {
        return false;
      }
      if (recordLineId) {
        return (record.LineId || "") === recordLineId;
      }
      return (record.Line || TENCENTCLOUD_DEFAULT_LINE) === recordLine;
    });

    if (existing) {
      if (existing.Value === ip) {
        return;
      }

      const result =
        await tencentcloudRequest<TencentCloudRecordChangeResponse>(
          context,
          "ModifyRecord",
          {
            ...basePayload,
            RecordId: existing.RecordId,
            SubDomain: parsed.recordName,
            TTL: ttl,
            Value: ip,
          },
        );

      if (!result.RecordId) {
        throw new Error(tencentcloudT("missingUpdatedRecordId"));
      }
      return;
    }

    const result = await tencentcloudRequest<TencentCloudRecordChangeResponse>(
      context,
      "CreateRecord",
      {
        ...basePayload,
        SubDomain: parsed.recordName,
        TTL: ttl,
        Value: ip,
      },
    );

    if (!result.RecordId) {
      throw new Error(tencentcloudT("missingCreatedRecordId"));
    }
  });
}

File: apps/server-admin-rs/src/ddns/routes/providers/tencentcloud.rs
use super::*;

pub(in crate::ddns::routes) fn tencentcloud_catalog_entry() -> Value {
    provider(
        "tencentcloud",
        "腾讯云 DNSPod",
        vec![
            field("secret_id", "SecretId", "text", "AKID...", true),
            field("secret_key", "SecretKey", "password", "SecretKey", true),
            field("root_domain", "Root Domain", "text", "example.com", true),
            field("domain", "Domain", "text", "home.example.com", true),
            field("record_line", "Record Line", "text", "默认", false),
            field("record_line_id", "Record Line ID", "text", "0", false),
            field("ttl", "TTL", "text", "600", false),
        ],
    )
}

pub(in crate::ddns::routes) async fn update_tencentcloud(
    translator: &Translator,
    config: &HashMap<String, String>,
    http_options: &DDNSHttpClientOptions,
    ipv4: Option<&str>,
    ipv6: Option<&str>,
) -> anyhow::Result<DDNSProviderUpdateResult> {
    let secret_id = config_value(config, "secret_id");
    let secret_key = config_value(config, "secret_key");
    let root_domain = config_value(config, "root_domain");
    let domain = config_value(config, "domain");
    if secret_id.is_empty() || secret_key.is_empty() || root_domain.is_empty() || domain.is_empty()
    {
        return Ok(provider_failure(ddns_text(
            translator,
            "providers.tencentcloud.configIncomplete",
            &[],
        )));
    }
    let ttl = positive_i64(config.get("ttl"), 600);
    let record_line = default_string(config_value(config, "record_line"), "默认");
    let record_line_id = config_value(config, "record_line_id");
    let parsed = split_domain(translator, &domain, &root_domain)?;
    let client = ddns_http_client(translator, http_options)?;
    let missing_updated_record_id = ddns_text(
        translator,
        "providers.tencentcloud.missingUpdatedRecordId",
        &[],
    );
    let missing_created_record_id = ddns_text(
        translator,
        "providers.tencentcloud.missingCreatedRecordId",
        &[],
    );

    let provider_label_text = provider_label(Some("tencentcloud"), translator);
    update_dual_stack(
        translator,
        &provider_label_text,
        ipv4,
        ipv6,
        |record_type, ip| {
            let client = client.clone();
            let secret_id = secret_id.clone();
            let secret_key = secret_key.clone();
            let parsed = parsed.clone();
            let record_line = record_line.clone();
            let record_line_id = record_line_id.clone();
            let missing_updated_record_id = missing_updated_record_id.clone();
            let missing_created_record_id = missing_created_record_id.clone();
            async move {
                let mut base_payload = serde_json::Map::new();
                base_payload.insert("Domain".to_string(), json!(parsed.root_domain));
                base_payload.insert("RecordType".to_string(), json!(record_type));
                if record_line_id.is_empty() {
                    base_payload.insert("RecordLine".to_string(), json!(record_line));
                } else {
                    base_payload.insert("RecordLineId".to_string(), json!(record_line_id));
                }

                let mut list_payload = base_payload.clone();
                list_payload.insert("Limit".to_string(), json!(100));
                list_payload.insert("Offset".to_string(), json!(0));
                list_payload.insert("Subdomain".to_string(), json!(parsed.record_name));
                let list = match tencentcloud_request(
                    translator,
                    &client,
                    &secret_id,
                    &secret_key,
                    "DescribeRecordList",
                    Value::Object(list_payload),
                )
                .await
                {
                    Ok(value) => value,
                    Err(error)
                        if error
                            .to_string()
                            .starts_with("ResourceNotFound.NoDataOfRecord:") =>
                    {
                        json!({ "RecordList": [] })
                    }
                    Err(error) => return Err(error),
                };
                let existing = list
                    .get("RecordList")
                    .and_then(Value::as_array)
                    .into_iter()
                    .flatten()
                    .find(|record| {
                        record
                            .get("Name")
                            .and_then(Value::as_str)
                            .unwrap_or(&parsed.record_name)
                            == parsed.record_name
                            && record
                                .get("Type")
                                .and_then(Value::as_str)
                                .unwrap_or(record_type)
                                == record_type
                            && if record_line_id.is_empty() {
                                record.get("Line").and_then(Value::as_str).unwrap_or("默认")
                                    == record_line
                            } else {
                                record
                                    .get("LineId")
                                    .and_then(Value::as_str)
                                    .unwrap_or_default()
                                    == record_line_id
                            }
                    })
                    .cloned();
                if let Some(record) = existing {
                    if record.get("Value").and_then(Value::as_str) == Some(ip.as_str()) {
                        return Ok(());
                    }
                    let record_id = record
                        .get("RecordId")
                        .filter(|value| json_value_js_truthy(Some(value)))
                        .cloned()
                        .ok_or_else(|| anyhow::anyhow!(missing_updated_record_id.clone()))?;
                    let mut payload = base_payload;
                    payload.insert("RecordId".to_string(), record_id);
                    payload.insert("SubDomain".to_string(), json!(parsed.record_name));
                    payload.insert("TTL".to_string(), json!(ttl));
                    payload.insert("Value".to_string(), json!(ip));
                    let result = tencentcloud_request(
                        translator,
                        &client,
                        &secret_id,
                        &secret_key,
                        "ModifyRecord",
                        Value::Object(payload),
                    )
                    .await?;
                    if json_value_js_truthy(result.get("RecordId")) {
                        return Ok(());
                    }
                    return Err(anyhow::anyhow!(missing_updated_record_id));
                }

                let mut payload = base_payload;
                payload.insert("SubDomain".to_string(), json!(parsed.record_name));
                payload.insert("TTL".to_string(), json!(ttl));
                payload.insert("Value".to_string(), json!(ip));
                let result = tencentcloud_request(
                    translator,
                    &client,
                    &secret_id,
                    &secret_key,
                    "CreateRecord",
                    Value::Object(payload),
                )
                .await?;
                if json_value_js_truthy(result.get("RecordId")) {
                    Ok(())
                } else {
                    Err(anyhow::anyhow!(missing_created_record_id))
                }
            }
        },
    )
    .await
}

pub(in crate::ddns::routes) async fn tencentcloud_request(
    translator: &Translator,
    client: &DDNSHttpClient,
    secret_id: &str,
    secret_key: &str,
    action: &str,
    payload: Value,
) -> anyhow::Result<Value> {
    const HOST: &str = "dnspod.tencentcloudapi.com";
    const SERVICE: &str = "dnspod";
    const VERSION: &str = "2021-03-23";

    let timestamp = time_utils::now_ms().div_euclid(1000);
    let date = utc_date(timestamp)?;
    let payload_string = serde_json::to_string(&payload)?;
    let hashed_payload = sha256_hex(&payload_string);
    let content_type = "application/json; charset=utf-8";
    let canonical_headers = tencentcloud_tc3_canonical_headers(content_type, HOST, action);
    let signed_headers = "content-type;host;x-tc-action";
    let canonical_request = [
        "POST",
        "/",
        "",
        &canonical_headers,
        signed_headers,
        &hashed_payload,
    ]
    .join("\n");
    let credential_scope = format!("{date}/{SERVICE}/tc3_request");
    let string_to_sign = [
        "TC3-HMAC-SHA256",
        &timestamp.to_string(),
        &credential_scope,
        &sha256_hex(&canonical_request),
    ]
    .join("\n");
    let secret_date = hmac_sha256_bytes(format!("TC3{secret_key}").as_bytes(), date.as_bytes());
    let secret_service = hmac_sha256_bytes(&secret_date, SERVICE.as_bytes());
    let secret_signing = hmac_sha256_bytes(&secret_service, b"tc3_request");
    let signature = hmac_sha256_hex(&secret_signing, string_to_sign.as_bytes());
    let authorization = format!(
        "TC3-HMAC-SHA256 Credential={secret_id}/{credential_scope}, SignedHeaders={signed_headers}, Signature={signature}"
    );

    let (status, data, _text) = response_json(
        translator,
        client
            .post(format!("https://{HOST}/"))
            .header(reqwest::header::CONTENT_TYPE, content_type)
            .header(reqwest::header::HOST, HOST)
            .header("X-TC-Action", action)
            .header("X-TC-Timestamp", timestamp.to_string())
            .header("X-TC-Version", VERSION)
            .header(reqwest::header::AUTHORIZATION, authorization)
            .body(payload_string)
            .send()
            .await?,
    )
    .await?;
    let response = data.get("Response").cloned().ok_or_else(|| {
        anyhow::anyhow!(ddns_text(
            translator,
            "tencentMissingResponse",
            &[("status", status.as_u16().to_string())],
        ))
    })?;
    if let Some(error) = response.get("Error") {
        let code = error
            .get("Code")
            .and_then(Value::as_str)
            .unwrap_or("TencentCloudError");
        let request_failed = ddns_text(translator, "requestFailed", &[]);
        let message = error
            .get("Message")
            .and_then(Value::as_str)
            .unwrap_or(request_failed.as_str());
        let request_id = response
            .get("RequestId")
            .and_then(Value::as_str)
            .unwrap_or_default();
        return Err(anyhow::anyhow!(
            "{code}: {message}{}",
            if request_id.is_empty() {
                String::new()
            } else {
                format!(" (RequestId: {request_id})")
            }
        ));
    }
    if status.is_success() {
        Ok(response)
    } else {
        Err(anyhow::anyhow!(
            "HTTP {}: {}",
            status.as_u16(),
            ddns_text(translator, "requestFailed", &[])
        ))
    }
}
